File: WMS.DALCore/Mensaje_regla/clsLnMensaje_regla.cs
using System.Data;
using Microsoft.Data.SqlClient;
using Microsoft.Extensions.Configuration;
using Microsoft.VisualBasic.CompilerServices;
using WMSWebAPI.Be;

public class clsLnMensaje_regla
{

    private static clsInsert Ins = new clsInsert();
    private static clsUpdate Upd = new clsUpdate();

    public static void Cargar(ref clsBeMensaje_regla oBeMensaje_regla, DataRow dr)
    {
        int GetInt(string col) { return dr[col] is DBNull ? 0 : Convert.ToInt32(dr[col]); }
        bool GetBool(string col) { return dr[col] is DBNull ? false : Convert.ToBoolean(dr[col]); }
        string GetString(string col) { return dr[col] is DBNull ? "" : (Convert.ToString(dr[col]) ?? ""); }
        DateTime GetDate(string col) { return dr[col] is DBNull ? DateTime.Now : Convert.ToDateTime(dr[col]); }

        try
        {
            oBeMensaje_regla.IdMensajeRegla = GetInt("IdMensajeRegla");
            oBeMensaje_regla.Nombre = GetString("Nombre");
            oBeMensaje_regla.Fec_agr = GetDate("fec_agr");
            oBeMensaje_regla.User_agr = GetString("user_agr");
            oBeMensaje_regla.Fec_mod = GetDate("fec_mod");
            oBeMensaje_regla.User_mod = GetString("user_mod");
            oBeMensaje_regla.Activo = GetBool("activo");
        }
        catch (Exception)
        {
            throw;
        }
    }

    public static int Insertar(clsBeMensaje_regla oBeMensaje_regla, SqlConnection pConection, SqlTransaction pTransaction)
    {
        int rowsAffected = 0;

        try
        {
            Ins.Init("mensaje_regla");
            Ins.Add("idmensajeregla", "@idmensajeregla", "F");
            Ins.Add("nombre", "@nombre", "F");
            Ins.Add("fec_agr", "@fec_agr", "F");
            Ins.Add("user_agr", "@user_agr", "F");
            Ins.Add("fec_mod", "@fec_mod", "F");
            Ins.Add("user_mod", "@user_mod", "F");
            Ins.Add("activo", "@activo", "F");

            string sp = Ins.SQL();

            using (var cmd = new SqlCommand(sp, pConection, pTransaction))
            {
                cmd.CommandType = CommandType.Text;
                cmd.Parameters.Add(new SqlParameter("@IdMensajeRegla", oBeMensaje_regla.IdMensajeRegla));
                cmd.Parameters.Add(new SqlParameter("@Nombre", oBeMensaje_regla.Nombre));
                cmd.Parameters.Add(new SqlParameter("@fec_agr", oBeMensaje_regla.Fec_agr));
                cmd.Parameters.Add(new SqlParameter("@user_agr", oBeMensaje_regla.User_agr));
                cmd.Parameters.Add(new SqlParameter("@fec_mod", oBeMensaje_regla.Fec_mod));
                cmd.Parameters.Add(new SqlParameter("@user_mod", oBeMensaje_regla.User_mod));
                cmd.Parameters.Add(new SqlParameter("@activo", oBeMensaje_regla.Activo));

                rowsAffected = cmd.ExecuteNonQuery();
            }

            return rowsAffected;
        }
        catch (SqlException)
        {
            throw;
        }
    }
    public static int Insertar(IConfiguration config, clsBeMensaje_regla oBeMensaje_regla)
    {

        int rowsAffected = 0;
        SqlConnection lConnection = new SqlConnection(config.GetConnectionString("CST"));
        SqlTransaction? lTransaction = null;

        try
        {
            Ins.Init("mensaje_regla");
            Ins.Add("idmensajeregla", "@idmensajeregla", "F");
            Ins.Add("nombre", "@nombre", "F");
            Ins.Add("fec_agr", "@fec_agr", "F");
            Ins.Add("user_agr", "@user_agr", "F");
            Ins.Add("fec_mod", "@fec_mod", "F");
            Ins.Add("user_mod", "@user_mod", "F");
            Ins.Add("activo", "@activo", "F");

            string sp = Ins.SQL();

            SqlCommand cmd = new SqlCommand() { CommandType = CommandType.Text };

            lConnection.Open(); lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUncommitted);
            cmd = new SqlCommand(sp, lConnection, lTransaction);

            cmd.Parameters.Add(new SqlParameter("@IdMensajeRegla", oBeMensaje_regla.IdMensajeRegla));
            cmd.Parameters.Add(new SqlParameter("@Nombre", oBeMensaje_regla.Nombre));
            cmd.Parameters.Add(new SqlParameter("@fec_agr", oBeMensaje_regla.Fec_agr));
            cmd.Parameters.Add(new SqlParameter("@user_agr", oBeMensaje_regla.User_agr));
            cmd.Parameters.Add(new SqlParameter("@fec_mod", oBeMensaje_regla.Fec_mod));
            cmd.Parameters.Add(new SqlParameter("@user_mod", oBeMensaje_regla.User_mod));
            cmd.Parameters.Add(new SqlParameter("@activo", oBeMensaje_regla.Activo));

            rowsAffected = cmd.ExecuteNonQuery();

            if (lTransaction != null)
                lTransaction.Commit();

        }
        catch (SqlException)
        {
            if (lTransaction is not null)
                lTransaction.Rollback();
            throw;
        }
        finally
        {
            if (lConnection.State == ConnectionState.Open) lConnection.Close();
            if (lConnection != null) lConnection.Dispose();
            if (lTransaction != null) lTransaction.Dispose();
        }
        return rowsAffected;
    }

    public static int Actualizar(clsBeMensaje_regla oBeMensaje_regla, SqlConnection pConection, SqlTransaction pTransaction)
    {
        int rowsAffected = 0;

        try
        {
            Upd.Init("mensaje_regla");
            Upd.Add("idmensajeregla", "@idmensajeregla", "F");
            Upd.Add("nombre", "@nombre", "F");
            Upd.Add("fec_agr", "@fec_agr", "F");
            Upd.Add("user_agr", "@user_agr", "F");
            Upd.Add("fec_mod", "@fec_mod", "F");
            Upd.Add("user_mod", "@user_mod", "F");
            Upd.Add("activo", "@activo", "F");
            Upd.Where("IdMensajeRegla = @IdMensajeRegla");

            string sp = Upd.SQL();

            using (SqlCommand cmd = new SqlCommand(sp, pConection, pTransaction))
            {
                cmd.CommandType = CommandType.Text;
                cmd.Parameters.Add(new SqlParameter("@IdMensajeRegla", oBeMensaje_regla.IdMensajeRegla));
                cmd.Parameters.Add(new SqlParameter("@Nombre", oBeMensaje_regla.Nombre));
                cmd.Parameters.Add(new SqlParameter("@fec_agr", oBeMensaje_regla.Fec_agr));
                cmd.Parameters.Add(new SqlParameter("@user_agr", oBeMensaje_regla.User_agr));
                cmd.Parameters.Add(new SqlParameter("@fec_mod", oBeMensaje_regla.Fec_mod));
                cmd.Parameters.Add(new SqlParameter("@user_mod", oBeMensaje_regla.User_mod));
                cmd.Parameters.Add(new SqlParameter("@activo", oBeMensaje_regla.Activo));

                rowsAffected = cmd.ExecuteNonQuery();
            }

            return rowsAffected;
        }
        catch (SqlException)
        {
            throw;
        }
    }

    public int Eliminar(IConfiguration config, clsBeMensaje_regla oBeMensaje_regla, SqlConnection? pConection = null, SqlTransaction? pTransaction = null)
    {

        SqlConnection lConnection = new SqlConnection(config.GetConnectionString("CST"));
        SqlTransaction? lTransaction = null;

        try
        {
            const string sp = (" Delete from Mensaje_regla" +
             "  Where(IdMensajeRegla = @IdMensajeRegla)");

            bool Es_Transaccion_Remota = (pConection != null && pTransaction != null);

            SqlCommand cmd = new SqlCommand() { CommandType = CommandType.Text };

            if (Es_Transaccion_Remota)
            {
                cmd = new SqlCommand(sp, pConection, pTransaction);
            }
            else
            {
                lConnection.Open(); lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUncommitted);
                cmd = new SqlCommand(sp, lConnection, lTransaction);
            }

            cmd.Parameters.Add(new SqlParameter("@IdMensajeRegla", oBeMensaje_regla.IdMensajeRegla));

            int rowsAffected = cmd.ExecuteNonQuery();

            if (!Es_Transaccion_Remota)
                if (lTransaction != null)
                    lTransaction.Commit();

            return rowsAffected;

        }
        catch (SqlException)
        {
            if (lTransaction is not null)
                lTransaction.Rollback();
            throw;
        }
        finally
        {
            if (lConnection.State == ConnectionState.Open) lConnection.Close();
            if (lConnection != null) lConnection.Dispose();
            if (lTransaction != null) lTransaction.Dispose();
        }
    }

    public DataTable Listar(IConfiguration config)
    {

        SqlConnection lConnection = new SqlConnection(config.GetConnectionString("CST"));
        SqlTransaction? lTransaction = null;

        try
        {
            const string sp = "Select * FROM Mensaje_regla";
            lConnection.Open(); lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUncommitted);
            SqlCommand cmd = new SqlCommand(sp, lConnection, lTransaction) { CommandType = CommandType.Text };
            SqlDataAdapter dad = new SqlDataAdapter(cmd);
            DataTable dt = new DataTable();
            dad.Fill(dt);

            lTransaction.Commit();

            return dt;

        }
        catch (SqlException)
        {
            if (lTransaction is not null)
                lTransaction.Rollback();
            throw;
        }
        finally
        {
            if (lConnection.State == ConnectionState.Open) lConnection.Close();
            if (lConnection != null) lConnection.Dispose();
            if (lTransaction != null) lTransaction.Dispose();
        }
    }

    public static bool GetSingle(IConfiguration config, ref clsBeMensaje_regla pBeMensaje_regla)
    {

        SqlConnection lConnection = new SqlConnection(config.GetConnectionString("CST"));
        SqlTransaction? lTransaction = null;

        try
        {

            const string sp = "Select * FROM Mensaje_regla" +
            " Where(IdMensajeRegla = @IdMensajeRegla)";

            lConnection.Open(); lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUncommitted);

            SqlCommand cmd = new SqlCommand(sp, lConnection, lTransaction) { CommandType = CommandType.Text };
            SqlDataAdapter dad = new SqlDataAdapter(cmd);

            dad.SelectCommand.Parameters.Add(new SqlParameter("@IdMensajeRegla", pBeMensaje_regla.IdMensajeRegla));
            dad.SelectCommand.Parameters.Add(new SqlParameter("@Nombre", pBeMensaje_regla.Nombre));
            dad.SelectCommand.Parameters.Add(new SqlParameter("@fec_agr", pBeMensaje_regla.Fec_agr));
            dad.SelectCommand.Parameters.Add(new SqlParameter("@user_agr", pBeMensaje_regla.User_agr));
            dad.SelectCommand.Parameters.Add(new SqlParameter("@fec_mod", pBeMensaje_regla.Fec_mod));
            dad.SelectCommand.Parameters.Add(new SqlParameter("@user_mod", pBeMensaje_regla.User_mod));
            dad.SelectCommand.Parameters.Add(new SqlParameter("@activo", pBeMensaje_regla.Activo));

            DataTable dt = new DataTable();
            dad.Fill(dt);

            lTransaction.Commit();

            if (dt.Rows.Count == 1)
            {
                DataRow r;
                r = dt.Rows[0];
                Cargar(ref pBeMensaje_regla, r);
                return true;
            }

        }
        catch (SqlException)
        {
            if (lTransaction is not null)
                lTransaction.Rollback();
            throw;
        }
        finally
        {
            if (lConnection.State == ConnectionState.Open) lConnection.Close();
            if (lConnection != null) lConnection.Dispose();
            if (lTransaction != null) lTransaction.Dispose();
        }
        return false;

    }

    public static List<clsBeMensaje_regla> GetAll(IConfiguration config)
    {

        SqlTransaction? lTransaction = null;
        List<clsBeMensaje_regla> lreturnList = new List<clsBeMensaje_regla>();

        try
        {
            const string sp = "Select * FROM Mensaje_regla";

            using (SqlConnection lConnection = new SqlConnection(config.GetConnectionString("CST")))
            {

                lConnection.Open();

                using (lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUncommitted))
                {
                    using (SqlDataAdapter lDTA = new SqlDataAdapter(sp, lConnection))
                    {
                        lDTA.SelectCommand.CommandType = CommandType.Text;
                        lDTA.SelectCommand.Transaction = lTransaction;
                        DataTable lDataTable = new DataTable();
                        lDTA.Fill(lDataTable);

                        clsBeMensaje_regla vBeMensaje_regla = new clsBeMensaje_regla();

                        foreach (DataRow dr in lDataTable.Rows)
                        {
                            vBeMensaje_regla = new clsBeMensaje_regla();
                            Cargar(ref vBeMensaje_regla, dr);
                            lreturnList.Add(vBeMensaje_regla);
                        }

                        lTransaction.Commit();
                    }

                    lConnection.Close();

                }

            }

            return lreturnList;

        }
        catch (SqlException)
        {
            if (lTransaction is not null)
                lTransaction.Rollback();
            throw;
        }
    }

    public static int MaxID(IConfiguration config)
    {

        SqlTransaction? lTransaction = null;

        try
        {

            int lMax = 0;

            const string sp = "Select ISNULL(Max(IdMensajeRegla),0) FROM Mensaje_regla";

            using (SqlConnection lConnection = new SqlConnection(config.GetConnectionString("CST")))
            {
                lConnection.Open();

                using (lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUncommitted))
                {
                    using (SqlCommand lCommand = new SqlCommand(sp, lConnection) { CommandType = CommandType.Text })
                    {
                        var lreturnValue = lCommand.ExecuteScalar();
                        if (lreturnValue != DBNull.Value && lreturnValue != null)
                        {
                            lMax = int.Parse((string)lreturnValue);
                        }
                    }
                    lTransaction.Commit();
                }

                lConnection.Close();
            }

            return lMax;

        }
        catch (SqlException)
        {
            if (lTransaction is not null)
                lTransaction.Rollback();
            throw;
        }
    }
    public static int MaxID(SqlConnection pConection, SqlTransaction pTransaction)
    {
        int lMax = 0;
        try
        {
            const string sp = "Select ISNULL(Max(IdMensajeRegla),0) FROM Mensaje_regla";

            using (var cmd = new SqlCommand(sp, pConection, pTransaction))
            {
                cmd.CommandType = CommandType.Text;
                object lreturnValue = cmd.ExecuteScalar();

                if (lreturnValue != DBNull.Value && lreturnValue != null)
                {
                    lMax = Convert.ToInt32(lreturnValue);
                }
            }

            return lMax;
        }
        catch (SqlException)
        {
            throw;
        }
    }

}

File: TOMIMSV4/DAL/Mantenimientos/Mensaje_Regla/clsLnMensaje_regla.vb
Imports System.Data.SqlClient
Imports System.Reflection

Public Class clsLnMensaje_regla

    Public Shared Sub Cargar(ByRef oBeMensaje_regla As clsBeMensaje_regla, ByRef dr As DataRow)
        Try
            With oBeMensaje_regla
                .IdMensajeRegla = IIf(IsDBNull(dr.Item("IdMensajeRegla")), 0, dr.Item("IdMensajeRegla"))
                .Nombre = IIf(IsDBNull(dr.Item("Nombre")), "", dr.Item("Nombre"))
                .Fec_agr = IIf(IsDBNull(dr.Item("fec_agr")), Date.Now, dr.Item("fec_agr"))
                .User_agr = IIf(IsDBNull(dr.Item("user_agr")), "", dr.Item("user_agr"))
                .Fec_mod = IIf(IsDBNull(dr.Item("fec_mod")), Date.Now, dr.Item("fec_mod"))
                .User_mod = IIf(IsDBNull(dr.Item("user_mod")), "", dr.Item("user_mod"))
                .Activo = IIf(IsDBNull(dr.Item("activo")), False, dr.Item("activo"))
            End With

        Catch ex1 As SQLException
            Throw ex1
        Catch ex As Exception
            Dim vMsgError As String = String.Format("{0} {1}", MethodBase.GetCurrentMethod.Name(), ex.Message)
            clsLnLog_error_wms.Agregar_Error(vMsgError)
            Throw ex
        End Try
    End Sub

    Public Shared Function Insertar(ByRef oBeMensaje_regla As clsBeMensaje_regla, Optional ByVal pConection as SqlConnection = Nothing, Optional Byval pTransaction as SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing

        Try

            Ins.Init("mensaje_regla")
            Ins.Add("idmensajeregla", "@idmensajeregla", DataType.Parametro)
            Ins.Add("nombre", "@nombre", DataType.Parametro)
            Ins.Add("fec_agr", "@fec_agr", DataType.Parametro)
            Ins.Add("user_agr", "@user_agr", DataType.Parametro)
            Ins.Add("fec_mod", "@fec_mod", DataType.Parametro)
            Ins.Add("user_mod", "@user_mod", DataType.Parametro)
            Ins.Add("activo", "@activo", DataType.Parametro)
            Ins.Add("IdreglaRecepcion", "@IdreglaRecepcion", DataType.Parametro)

            Dim sp As String = Ins.SQL()
            Dim cmd As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}

            Dim Es_Transaccion_Remota As Boolean = (pConection IsNot Nothing AndAlso pTransaction IsNot Nothing)

            If Es_Transaccion_Remota Then
                cmd = New SqlCommand(sp, pConection)
                cmd.Transaction = pTransaction
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)
            End If

            cmd.Parameters.Add(New SqlParameter("@IDMENSAJEREGLA", oBeMensaje_regla.IdMensajeRegla))
            cmd.Parameters.Add(New SqlParameter("@NOMBRE", oBeMensaje_regla.Nombre))
            cmd.Parameters.Add(New SqlParameter("@FEC_AGR", oBeMensaje_regla.Fec_agr))
            cmd.Parameters.Add(New SqlParameter("@USER_AGR", oBeMensaje_regla.User_agr))
            cmd.Parameters.Add(New SqlParameter("@FEC_MOD", oBeMensaje_regla.Fec_mod))
            cmd.Parameters.Add(New SqlParameter("@USER_MOD", oBeMensaje_regla.User_mod))
            cmd.Parameters.Add(New SqlParameter("@ACTIVO", oBeMensaje_regla.Activo))
            cmd.Parameters.Add(New SqlParameter("@IDREGLARECEPCION", oBeMensaje_regla.IdReglaRecepcion))

            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()

            cmd.Dispose()

            If Not Es_Transaccion_Remota Then lTransaction.Commit()

            Return rowsAffected

        Catch ex1 As SqlException
            If lTransaction IsNot Nothing Then lTransaction.Rollback()
            Throw ex1
        Catch ex As Exception
            If lTransaction IsNot Nothing Then lTransaction.Rollback()
            Throw ex
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close()
            If lTransaction IsNot Nothing Then lTransaction.Dispose()
            If lConnection IsNot Nothing Then lConnection.Dispose()
        End Try

    End Function

    Public Shared Function Actualizar(ByRef oBeMensaje_regla As clsBeMensaje_regla, Optional ByVal pConection As SqlConnection = Nothing, Optional ByVal pTransaction As SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing

        Try

            Upd.Init("mensaje_regla")
            Upd.Add("idmensajeregla", "@idmensajeregla", DataType.Parametro)
            Upd.Add("nombre", "@nombre", DataType.Parametro)
            Upd.Add("fec_agr", "@fec_agr", DataType.Parametro)
            Upd.Add("user_agr", "@user_agr", DataType.Parametro)
            Upd.Add("fec_mod", "@fec_mod", DataType.Parametro)
            Upd.Add("user_mod", "@user_mod", DataType.Parametro)
            Upd.Add("activo", "@activo", DataType.Parametro)
            Upd.Where("IdMensajeRegla = @IdMensajeRegla")

            Dim sp As String = Upd.SQL()

            Dim Es_Transaccion_Remota As Boolean = (pConection IsNot Nothing AndAlso pTransaction IsNot Nothing)
            Dim cmd As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}


            If Es_Transaccion_Remota Then
                cmd = New SqlCommand(sp, pConection)
                cmd.Transaction = pTransaction
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)
            End If

            cmd.Parameters.Add(New SqlParameter("@IDMENSAJEREGLA", oBeMensaje_regla.IdMensajeRegla))
            cmd.Parameters.Add(New SqlParameter("@NOMBRE", oBeMensaje_regla.Nombre))
            cmd.Parameters.Add(New SqlParameter("@FEC_AGR", oBeMensaje_regla.Fec_agr))
            cmd.Parameters.Add(New SqlParameter("@USER_AGR", oBeMensaje_regla.User_agr))
            cmd.Parameters.Add(New SqlParameter("@FEC_MOD", oBeMensaje_regla.Fec_mod))
            cmd.Parameters.Add(New SqlParameter("@USER_MOD", oBeMensaje_regla.User_mod))
            cmd.Parameters.Add(New SqlParameter("@ACTIVO", oBeMensaje_regla.Activo))

            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()

            cmd.Dispose()

            If Not Es_Transaccion_Remota Then lTransaction.Commit()

            Return rowsAffected

        Catch ex1 As SqlException
            If lTransaction IsNot Nothing Then lTransaction.Rollback()
            Throw ex1
        Catch ex As Exception
            If lTransaction IsNot Nothing Then lTransaction.Rollback()
            Throw ex
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close()
            If lTransaction IsNot Nothing Then lTransaction.Dispose()
            If lConnection IsNot Nothing Then lConnection.Dispose()
        End Try

    End Function


    Public Shared Function Eliminar(ByRef oBeMensaje_regla As clsBeMensaje_regla, Optional ByVal pConection As SqlConnection = Nothing, Optional ByVal pTransaction As SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing

        Try

            Const sp As String = " Delete from Mensaje_regla" &
             "  Where(IdMensajeRegla = @IdMensajeRegla)"

            Dim Es_Transaccion_Remota As Boolean = (pConection IsNot Nothing AndAlso pTransaction IsNot Nothing)
            Dim cmd As New SqlCommand With {.CommandType = CommandType.Text}

            If Es_Transaccion_Remota Then
                cmd = New SqlCommand(sp, pConection)
                cmd.Transaction = pTransaction
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)
            End If

            cmd.Parameters.Add(New SqlParameter("@IDMENSAJEREGLA", oBeMensaje_regla.IdMensajeRegla))

            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()

            cmd.Dispose()

            If Not Es_Transaccion_Remota Then lTransaction.Commit()

            Return rowsAffected

        Catch ex1 As SqlException
            If lTransaction IsNot Nothing Then lTransaction.Rollback()
            Throw ex1
        Catch ex As Exception
            If lTransaction IsNot Nothing Then lTransaction.Rollback()
            Throw ex
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close()
            If lTransaction IsNot Nothing Then lTransaction.Dispose()
            If lConnection IsNot Nothing Then lConnection.Dispose()
        End Try

    End Function

    Public Shared Function EliminarTodos(Optional ByVal pConection As SqlConnection = Nothing, Optional ByVal pTransaction As SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing

        Try

            Const sp As String = " Delete from Mensaje_regla"
            Dim cmd As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}
            Dim Es_Transaccion_Remota As Boolean = (pConection IsNot Nothing AndAlso pTransaction IsNot Nothing)

            If Es_Transaccion_Remota Then
                cmd = New SqlCommand(sp, pConection)
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)
            End If

            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()

            cmd.Dispose()

            If Not Es_Transaccion_Remota Then lTransaction.Commit()

            Return rowsAffected

        Catch ex1 As SqlException
            Throw ex1
        Catch ex As Exception
            Throw ex
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close
        End Try

    End Function

    Public Shared Function Listar() As DataTable

        Try

            Const sp As String = "SELECT * FROM Mensaje_regla"
            Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
            Dim cmd As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}
            Dim dad As New SqlDataAdapter(cmd)
            Dim dt As New DataTable
            dad.Fill(dt)

            Return dt


        Catch ex1 As SQLException
            Throw ex1
        Catch ex As Exception
            Dim vMsgError As String = String.Format("{0} {1}", MethodBase.GetCurrentMethod.Name(), ex.Message)
            clsLnLog_error_wms.Agregar_Error(vMsgError)
            Throw ex
        End Try

    End Function

    Public Shared Function Obtener(ByRef oBeMensaje_regla As clsBeMensaje_regla) As Boolean

        Try

            Const sp As String = "SELECT * FROM Mensaje_regla" & _
            " Where(IdMensajeRegla = @IdMensajeRegla)"

            Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
            Dim cmd As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}
            Dim dad As New SqlDataAdapter(cmd)


            dad.SelectCommand.Parameters.Add(New SqlParameter("@IDMENSAJEREGLA", oBeMensaje_regla.IDMENSAJEREGLA))

            Dim dt As New DataTable
            dad.Fill(dt)

            If dt.Rows.Count = 1 Then
                Cargar(oBeMensaje_regla, dt.Rows(0))
            Else
                Throw New Exception("No se pudo obtener el registro")
            End If

            Return True


        Catch ex1 As SQLException
            Throw ex1
        Catch ex As Exception
            Dim vMsgError As String = String.Format("{0} {1}", MethodBase.GetCurrentMethod.Name(), ex.Message)
            clsLnLog_error_wms.Agregar_Error(vMsgError)
            Throw ex
        End Try

    End Function

    Public Shared Function GetAll() As List(Of clsBeMensaje_regla)

        Try

            Dim lReturnList As New List(Of clsBeMensaje_regla)
            Const sp As String = "SELECT * FROM Mensaje_regla"
            Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
            Dim cmd As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}
            Dim dad As New SqlDataAdapter(cmd)
            Dim dt As New DataTable

            dad.Fill(dt)

            Dim vBeMensaje_regla As New clsBeMensaje_regla

            For Each dr As DataRow In dt.Rows

                vBeMensaje_regla = New clsBeMensaje_regla
                Cargar(vBeMensaje_regla, dr)
                lReturnList.Add(vBeMensaje_regla)

            Next

            lConnection.Dispose()
            cmd.Dispose()

            Return lReturnList

        Catch ex As Exception
            Throw New Exception(String.Format("{0} {1}", MethodBase.GetCurrentMethod.Name(), ex.Message))
        End Try

    End Function

    Public Shared Function GetSingle(ByRef pBeMensaje_regla As clsBeMensaje_regla)

        Try

            Const sp As String = "SELECT * FROM Mensaje_regla" & _
            " Where(IdMensajeRegla = @IdMensajeRegla)"

            Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
            Dim cmd As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}
            Dim dad As New SqlDataAdapter(cmd)
            dad.SelectCommand.Parameters.Add(New SqlParameter("@IDMENSAJEREGLA", pBeMensaje_regla.IDMENSAJEREGLA))

            Dim dt As New DataTable
            dad.Fill(dt)

            If dt.Rows.Count = 1 Then
                Cargar(pBeMensaje_regla, dt.Rows(0))
            End If

            Return True


        Catch ex1 As SQLException
            Throw ex1
        Catch ex As Exception
            Dim vMsgError As String = String.Format("{0} {1}", MethodBase.GetCurrentMethod.Name(), ex.Message)
            clsLnLog_error_wms.Agregar_Error(vMsgError)
            Throw ex
        End Try

    End Function

    Public Shared Function MaxID() as Integer

        Try

            Dim lMax As Integer = 0

            Const sp As String = "SELECT ISNULL(Max(IdMensajeRegla),0) FROM Mensaje_regla"

            Using lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
                lConnection.Open()
                Using lCommand As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}
                    Dim lReturnValue As Object = lCommand.ExecuteScalar()
                    lConnection.Close()
                    If lReturnValue IsNot DBNull.Value AndAlso lReturnValue IsNot Nothing Then
                        lMax = CInt(lReturnValue) + 1
                    End If
                End Using
            End Using

            Return lMax


        Catch ex1 As SQLException
            Throw ex1
        Catch ex As Exception
            Dim vMsgError As String = String.Format("{0} {1}", MethodBase.GetCurrentMethod.Name(), ex.Message)
            clsLnLog_error_wms.Agregar_Error(vMsgError)
            Throw ex
        End Try

    End Function

    Public Shared Function GetSingle(ByRef pBeMensaje_regla As clsBeMensaje_regla,
                                     ByVal lConnection As SqlConnection,
                                     ByVal lTransaction As SqlTransaction)

        Try

            Const sp As String = "SELECT * FROM Mensaje_regla" &
            " Where(IdMensajeRegla = @IdMensajeRegla)"
            Dim cmd As New SqlCommand(sp, lConnection, lTransaction) With {.CommandType = CommandType.Text}
            Dim dad As New SqlDataAdapter(cmd)
            dad.SelectCommand.Parameters.Add(New SqlParameter("@IDMENSAJEREGLA", pBeMensaje_regla.IdMensajeRegla))

            Dim dt As New DataTable
            dad.Fill(dt)

            If dt.Rows.Count = 1 Then
                Cargar(pBeMensaje_regla, dt.Rows(0))
            End If

            Return True

        Catch ex1 As SqlException
            Throw ex1
        Catch ex As Exception
            Dim vMsgError As String = String.Format("{0} {1}", MethodBase.GetCurrentMethod.Name(), ex.Message)
            clsLnLog_error_wms.Agregar_Error(vMsgError)
            Throw ex
        End Try

    End Function

End Class
